### PR TITLE
example using hf datasets

### PR DESCRIPTION
**Pull request by MaximeSongIdris**
### Checks to be done before submitting your PR
- [ ] `python3 -m pytest tests/` runs successfully.
- [ ] `black .` runs successfully.
- [ ] `make html` runs successfully (in the `docs/` directory).
- [ ] Updated docstrings related to the changes (as applicable).
- [ ] Added an entry to the [CHANGELOG.rst](../CHANGELOG.rst).

### CODE REVIEW

File: examples/basics/demo_hf_dataset.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+from torchvision import transforms
+
+import deepinv as dinv
+
+
+# %%
+# Stream data from Internet
+# ----------------------------------------------------------------------------------------
+#
+
```
Comment (Andrewwango) [inline examples/basics/demo_hf_dataset.py:28]: Remove this newline to render the comment as plaintext. Then make it more readable: 

> Stream data from HuggingFace. Only a limited number of samples should be in memory and nothing should be saved on disk. The dataset is downloaded from https://huggingface.co/datasets/deepinv/drunet_dataset. ``type : datasets.iterable_dataset.IterableDataset``

Screenshot 2024-07-21 235950

File: examples/basics/demo_hf_dataset.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+# stream data from huggingface
+# only a limited number of samples should be in memory and nothing should be saved on disk
+# https://huggingface.co/datasets/deepinv/drunet_dataset
+# type : datasets.iterable_dataset.IterableDataset
+raw_hf_train_dataset = load_dataset(
+    "deepinv/drunet_dataset", split="train", streaming=True
+)
+print("Number of data files used to store raw data: ", raw_hf_train_dataset.n_shards)
+
```
Comment (Andrewwango) [inline examples/basics/demo_hf_dataset.py:37]: Add newline then `# %%` to render a new cell for better readability.

> In streaming mode, we can only sequentially read the data sample in a certain order, thus we are not able to do exact shuffling. An alternative way is buffer shuffling, which loads a fixed number of samples in memory and lets us pick randomly one sample among this fixed number of samples:

File: examples/basics/demo_hf_dataset.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+# and let us pick randomly one sample among this fixed number of samples
+# https://huggingface.co/docs/datasets/about_mapstyle_vs_iterable
+raw_hf_train_dataset = raw_hf_train_dataset.shuffle(seed=42, buffer_size=100)
+
+
+# %%
+# Apply transformation on dataset
+# ----------------------------------------------------------------------------------------
+#
+# We define transformation with `torchvision.transforms` module.
```
Comment (Andrewwango) [inline examples/basics/demo_hf_dataset.py:50]: ```torchvision.transforms```

File: examples/basics/demo_hf_dataset.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+)
+
+
+# Class that apply transform on data samples of a datasets.iterable_dataset.IterableDataset
+class HFDataset(IterableDataset):
+    r"""
+    Creates an iteratble dataset from a Hugging Face dataset to enable streaming.
+    """
+
+    def __init__(self, hf_dataset, transforms=None, key="png"):
```
Comment (Andrewwango) [inline examples/basics/demo_hf_dataset.py:69]: `transform`

File: examples/basics/demo_hf_dataset.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+hf_train_dataset = HFDataset(raw_hf_train_dataset, transforms=img_transforms)
+
+
+# %%
+# Create a dataloader
+# --------------------------------------------------------------------
+#
+# We iterate over the dataset using a dataloader.
+#
+
```
Comment (Andrewwango) [inline examples/basics/demo_hf_dataset.py:92]: remove this newline and improve the readability of the below text

**Review by Andrewwango — CHANGES_REQUESTED** [review]

File: .github/workflows/documentation.yml [hunk 6]
```diff
@@ -27,6 +27,9 @@ jobs:
       - name: Install git repository optional dependencies
         run: |
           python3 -m pip install https://github.com/odlgroup/odl/archive/master.zip
+      - name: Install dependencies needed to run the HuggingFace dataset example
```
Comment (tachella) [inline .github/workflows/documentation.yml:30]: shouldn't this go in the optional dependencies?

Comment (MaximeSongIdris) [reply .github/workflows/documentation.yml:30]: yes, it was old code, removed lines

**Review by tachella — APPROVED** [review]
